File: src/mitris/sim/core/modeling/Coupling.java
package mitris.sim.core.modeling;

import mitris.sim.core.modeling.api.Port;


/**
 * TODO: Saurabh wants and interface for Coupling. I find no reason to create such interface. There is only
 * one kind of Coupling (for now), so no interfaces are needed.
 * @author José Luis Risco Martín
 */
public class Coupling<E> {

	protected Port<E> portFrom;
	protected Port<E> portTo;

	public Coupling(Port<E> portFrom, Port<E> portTo) {
		this.portFrom = portFrom;
		this.portTo = portTo;
	}

	public String toString(){
		return "(" + portFrom + "->" + portTo+")";
	}

	// Coupling members
	
	public void propagateValues() {
		portTo.addValues(portFrom.getValues());
	}

	public Port<E> getPortFrom() {
		return portFrom;
	}

	public Port<E> getPortTo() {
		return portTo;
	}
}

File: modeling/coupling.go
package modeling

import "reflect"

type Coupling interface {
	String() string
	PropagateValues()
	GetPortFrom() Port
	GetPortTo() Port
}

func NewCoupling(portFrom Port, portTo Port) Coupling {
	if reflect.TypeOf(portFrom.GetValues()) != reflect.TypeOf(portTo.GetValues()) {
		panic("Both ports of a coupling must be of the same type")
	}
	c := coupling{portFrom, portTo}
	return &c
}

type coupling struct {
	portFrom Port
	portTo Port
}

func (c *coupling) String() string {
	return "(" + c.portFrom.GetName() + "->" + c.portTo.GetName() + ")"
}

func (c *coupling) PropagateValues() {
	c.portTo.AddValues(c.portFrom.GetValues())
}

func (c *coupling) GetPortFrom() Port {
	return c.portFrom
}

func (c *coupling) GetPortTo() Port {
	return c.portTo
}
